db/db.go

```diff
@@ -208,38 +203,51 @@ func GetEth1DepositsJoinEth2Deposits(query string, length, start uint64, orderBy
 	var totalCount uint64
 	var err error
 
-	query = strings.Replace(query, "0x", "", -1)
+	deposistsCountQuery := `
+		SELECT COUNT(*) FROM eth1_deposits as eth1
+		%s`
 
-	if searchLikeHash.MatchString(query) {
-		if query != "" {
-			err = ReaderDb.Get(&totalCount, `
-				SELECT COUNT(*) FROM eth1_deposits as eth1
-				WHERE 
-					ENCODE(eth1.publickey, 'hex') LIKE LOWER($1)
-					OR ENCODE(eth1.withdrawal_credentials, 'hex') LIKE LOWER($1)
-					OR ENCODE(eth1.from_address, 'hex') LIKE LOWER($1)
-					OR ENCODE(tx_hash, 'hex') LIKE LOWER($1)
-					OR CAST(eth1.block_number AS text) LIKE LOWER($1)`, query+"%")
-		}
-	} else {
-		if query != "" {
-			err = ReaderDb.Get(&totalCount, `
-				SELECT COUNT(*) FROM eth1_deposits as eth1
-				WHERE 
-				CAST(eth1.block_number AS text) LIKE LOWER($1)`, query+"%")
+	trimmedQuery := strings.ToLower(strings.TrimPrefix(query, "0x"))
+	var hash []byte
+	if len(trimmedQuery)%2 == 0 && utils.HashLikeRegex.MatchString(trimmedQuery) {
+		hash, err = hex.DecodeString(trimmedQuery)
+		if err != nil {
+			return nil, 0, err
 		}
 	}
 
-	if query == "" {
-		err = ReaderDb.Get(&totalCount, "SELECT COUNT(*) FROM eth1_deposits")
+	if trimmedQuery != "" {
+		if utils.IsHash(trimmedQuery) {
+			searchQuery := `WHERE eth1.publickey = $1`
+			err = ReaderDb.Get(&totalCount, fmt.Sprintf(deposistsCountQuery, searchQuery), hash)
+		} else if utils.IsEth1Tx(trimmedQuery) {
+			// Withdrawal credentials have the same length as a tx hash
+			searchQuery := ""
+			if utils.IsValidWithdrawalCredentials(trimmedQuery) {
+				searchQuery = `
+				WHERE 
+					eth1.tx_hash = $1
+					OR eth1.withdrawal_credentials = $1`
+			} else {
+				searchQuery = `WHERE eth1.tx_hash = $1`
+			}
+			err = ReaderDb.Get(&totalCount, fmt.Sprintf(deposistsCountQuery, searchQuery), hash)
+		} else if utils.IsEth1Address(trimmedQuery) {
+			searchQuery := `WHERE eth1.from_address = $1`
+			err = ReaderDb.Get(&totalCount, fmt.Sprintf(deposistsCountQuery, searchQuery), hash)
+		} else if uiQuery, parseErr := strconv.ParseUint(query, 10, 64); parseErr == nil {
+			searchQuery := `WHERE eth1.block_number = $1`
+			err = ReaderDb.Get(&totalCount, fmt.Sprintf(deposistsCountQuery, searchQuery), uiQuery)
+		}
+	} else {
+		err = ReaderDb.Get(&totalCount, fmt.Sprintf(deposistsCountQuery, ""))
 	}
```

Comment:
It's not quite that easy because you don't want to do a query if `query` is not empty and does not fulfill at least one of the requirements.
You also can't have extra parameters like `param` if the query text has no `$1`.

I still rewrote those parts and for this function I used the same code for both queries by replacing `$3` => `$1` which seemed readable enough to me but please be the judge of that.